src/test/tst_attributerange.cpp

```diff
@@ -30,10 +31,632 @@ class TestAttributeRange: public QObject
+
+    // Tests if functions work as expected
+    QVERIFY(attrRge->isValid());
+    QCOMPARE(attrRge->id(), 0);
+    QCOMPARE(attrRge->attrName(), attrNames.at(0));
+    QCOMPARE(attrRge->attrRangeStr(), attrRangeStrs.at(0));
+
+    QVERIFY(attrRge2->isValid());
+    QCOMPARE(attrRge2->id(), 1);
+    QCOMPARE(attrRge2->attrName(), attrNames.at(1));
+    QCOMPARE(attrRge2->attrRangeStr(), attrRangeStrs.at(1));
+
+    QVERIFY(attrRge3->isValid());
+    QCOMPARE(attrRge3->id(), 2);
+    QCOMPARE(attrRge3->attrName(), attrNames.at(2));
+    QCOMPARE(attrRge3->attrRangeStr(), attrRangeStrs.at(2));
+
+    AttributeRange::Type type = AttributeRange::Double_Range;
+    QCOMPARE(attrRge->type(), type);
+    QCOMPARE(attrRge2->type(), type);
+    QCOMPARE(attrRge3->type(), type);
+
+    // Tests min(), max() and rand() functions
+    double min, max;
+    PRG* prg = new PRG(123);
+
+    min = 1.1;
+    max = 1.2;
+    QCOMPARE(attrRge->min().toDouble(), min);
+    QCOMPARE(attrRge->max().toDouble(), max);
+    v = attrRge->rand(prg);
+    _tst_value(v, Value::DOUBLE);
+    QVERIFY(v.toDouble() >= min);
+    QVERIFY(v.toDouble() <= max);
+
+    min = -5.5;
+    max = -3.3;
+    QCOMPARE(attrRge2->min().toDouble(), min);
+    QCOMPARE(attrRge2->max().toDouble(), max);
+    v = attrRge2->rand(prg);
+    _tst_value(v, Value::DOUBLE);
+    QVERIFY(v.toDouble() >= min);
+    QVERIFY(v.toDouble() <= max);
+
+    min = -95.7;
+    max = 87.5;
+    QCOMPARE(attrRge3->min().toDouble(), min);
+    QCOMPARE(attrRge3->max().toDouble(), max);
+    v = attrRge3->rand(prg);
+    _tst_value(v, Value::DOUBLE);
+    QVERIFY(v.toDouble() >= min);
+    QVERIFY(v.toDouble() <= max);
+
+    delete prg;
+}
+
+void TestAttributeRange::tst_int_set()
+{
+    const char* attrName = "test";
+    // Set with positive, negative and large numbers
+    const char* attrRangeStr = "int{0,1,-5,100,-100}";
+    int min = -100;
+    int max = 100;
+
+    AttributeRange* attrRge = AttributeRange::parse(0, attrName, attrRangeStr);
+
+    // Tests value returned by 'AttributeRange::validate()'
+    Value v;
+
+    v = attrRge->validate("0");
+    _tst_value(v, Value::INT);
+    v = attrRge->validate("1");
+    _tst_value(v, Value::INT);
+    v = attrRge->validate("-5");
+    _tst_value(v, Value::INT);
+    v = attrRge->validate("100");
+    _tst_value(v, Value::INT);
+    v = attrRge->validate("-100");
+    _tst_value(v, Value::INT);
+    v = attrRge->validate("invalid");  // invalid
+    _tst_value(v, Value::INVALID);
+
+    // Tests if functions work as expected
+    QVERIFY(attrRge->isValid());
+    QCOMPARE(attrRge->id(), 0);
+    QCOMPARE(attrRge->attrName(), QString(attrName));
+    QCOMPARE(attrRge->attrRangeStr(), QString(attrRangeStr));
+
+    AttributeRange::Type type = AttributeRange::Int_Set;
+    QCOMPARE(attrRge->type(), type);
+
+    // Tests min(), max() and rand() functions
+    PRG* prg = new PRG(123);
+
+    // min() returns the first value in the set
```

Comment:
very interesting! :clap: 

I hadn't considered this case before. I'd say that in an unordered set, it'd make more sense if `min()` always returns the smallest element in the set and not the first one. So, here, `min()` should return `-100`. I think that the only case that returning the first as 'min()' makes sense is for a set of strings.

Do you agree?

Reply:
Thank you, I would agree. If you happen to make that change in the future, I would be happy to test it!